otaclient/app/boot_control/firmware.py

```diff
@@ -0,0 +1,78 @@
+# Copyright 2023 TIER IV, INC. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import yaml
+import zstandard
+from pathlib import Path
+from typing import Dict, Callable
+from ..configs import config as cfg
+from .. import log_setting
+
+logger = log_setting.get_logger(
+    __name__, cfg.LOG_LEVEL_TABLE.get(__name__, cfg.DEFAULT_LOG_LEVEL)
+)
+
+
+class Firmware:
+    """
+    config_file example:
+    - `file` is expected to be located under the `config_file` directory.
+    - only zstd is support for now as compression.
+    - file is expected be extracted and copied to the appropriate partition.
+    - partitions[0] is for slot_a and partitions[1] is for slot_b
+    ---
+    version: 1
+    firmwares:
+      - file: rce.zst
+        compression: zstd
+        partitions:
+          - /dev/mmcblk0p21
+          - /dev/mmcblk0p22
+      - file: xusb.zst
+        compression: zstd
+        partitions:
+          - /dev/mmcblk0p19
+          - /dev/mmcblk0p20
+      ...
+    """
+
+    def __init__(self, config_file: Path):
+        self._config_file: Path = config_file
+
+    def update(self, slot_a: bool):
+        if not self._config_file.is_file():
+            # just emit warning for backward compatibility.
+            logger.warning(f"{self._config_file} doesn't exist")
+            return
```

Comment:
I think not updating firmware is the normal case(we only update firmware when BSP version migrates), so I think in contrast we should issue warning when we are going to update firmware.

Reply:
firmware updating is used only for the cboot and only cboot calls this update function, so I think it is natural to emit warning when config(=yaml) doesn't exist.